File: src/leetcode/binary_search/2300_successful_pairs_of_spells_and_potions.rb
# frozen_string_literal: true

# LeetCode 75
# https://leetcode.com/problems/successful-pairs-of-spells-and-potions/description/?envType=study-plan-v2&envId=leetcode-75

# Time complexity: O(log n)
# Space complexity: O(1)

# @param {Integer[]} spells
# @param {Integer[]} potions
# @param {Integer} success
# @return {Integer[]}
def successful_pairs(spells, potions, success)
  sorted_potions = potions.sort
  potion_length = sorted_potions.length
  max_potion = sorted_potions[potion_length - 1]

  spells.map do |spell|
    min_potion = (success / spell.to_f).ceil

    return 0 if min_potion > max_potion

    min_potion_index = binary_search(sorted_potions, min_potion)

    potion_length - min_potion_index
  end
end

def binary_search(potions, target_potion)
  low = 0
  high = potions.length

  while low < high
    mid = low + ((high - low) / 2)

    if potions[mid] < target_potion
      low = mid + 1
    else
      high = mid
    end
  end

  low
end

File: src/leetcode/binary_search/2300_successful_pairs_of_spells_and_potions.ts
// LeetCode 75
// https://leetcode.com/problems/successful-pairs-of-spells-and-potions/description/?envType=study-plan-v2&envId=leetcode-75

// Time complexity: O(log n)
// Space complexity: O(1)

const successfulPairs = (spells: number[], potions: number[], success: number): number[] => {
  const sortedPotions = potions.sort((a, b) => a - b);
  const potionsLength = sortedPotions.length;
  const maxPotion = sortedPotions[potionsLength - 1];

  return spells.map((spell) => {
    const minPotion = Math.ceil(success / spell);

    if (minPotion > maxPotion) {
      return 0;
    }

    const minPotionIndex = binary_search(sortedPotions, minPotion);

    return potionsLength - minPotionIndex;
  });
};

const binary_search = (potions: number[], targetPotion: number): number => {
  let low = 0;
  let high = potions.length;

  while (low < high) {
    const mid = Math.floor(low + (high - low) / 2);

    if (potions[mid] < targetPotion) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }

  return low;
};

export default successfulPairs;
